File: ios/Sources/AurigraphSDK/Errors/AurigraphError.swift
import Foundation

/// Errors thrown by the Aurigraph SDK.
public enum AurigraphError: Error, Equatable, Sendable {
    /// A network-level error (DNS, timeout, connection refused).
    case network(String)

    /// The server returned a non-2xx status code.
    /// If the body was RFC 7807 `application/problem+json`, `problem` is populated.
    case server(status: Int, message: String, problem: ProblemDetails?)

    /// JSON decoding failed.
    case decoding(String)

    /// Client misconfiguration (e.g. missing baseURL).
    case configuration(String)

    // MARK: - Equatable

    public static func == (lhs: AurigraphError, rhs: AurigraphError) -> Bool {
        switch (lhs, rhs) {
        case let (.network(a), .network(b)):
            return a == b
        case let (.server(s1, m1, _), .server(s2, m2, _)):
            return s1 == s2 && m1 == m2
        case let (.decoding(a), .decoding(b)):
            return a == b
        case let (.configuration(a), .configuration(b)):
            return a == b
        default:
            return false
        }
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/errors/AurigraphError.kt
package io.aurigraph.sdk.errors

import io.aurigraph.sdk.models.ProblemDetails

/**
 * Sealed exception hierarchy for all errors raised by the Aurigraph SDK.
 *
 * Uses RFC 7807 [ProblemDetails] when available from the server.
 */
sealed class AurigraphError(message: String, cause: Throwable? = null) : Exception(message, cause) {

    /**
     * HTTP 4xx client error with an RFC 7807 problem body.
     *
     * @property problem Parsed problem details from the server response.
     */
    class ClientError(val problem: ProblemDetails) :
        AurigraphError(problem.detail ?: problem.title ?: "Client error (${problem.status})")

    /**
     * HTTP 5xx server error with an RFC 7807 problem body.
     *
     * @property problem Parsed problem details from the server response.
     */
    class ServerError(val problem: ProblemDetails) :
        AurigraphError(problem.detail ?: problem.title ?: "Server error (${problem.status})")

    /**
     * Network-level failure (connection refused, timeout, DNS, etc.).
     */
    class NetworkError(cause: Throwable) :
        AurigraphError("Network error: ${cause.message}", cause)

    /**
     * SDK configuration error (missing baseUrl, invalid parameters, etc.).
     */
    class ConfigError(message: String) : AurigraphError(message)
}
